### ISSUE TITLE
Implement rule functions

### ISSUE THREAD

**Issue by NiklasRosenstein** [craftr-build/craftr#14]
The `compiler` modules should get rule functions that will generate the correct commands to build object files, libraries and executables.

```python
C = load_module('compiler')
build_dir = join(project_dir, 'build')

C.cc_objects('Objects',
  sources = glob(join(project_dir, 'src/**/*.c')),
  includes = [join(project_dir, 'include')],
)

C.cc_executable('Bin',
  name = 'main',
  inputs = Objects,
)
```

**Comment by winksaville** [comment]
How about passing a second parameter to load_module('compiler', 'c') then
C.executable(...) And C.objects(...)

**Comment by NiklasRosenstein** [comment]
There might be data in the module that you would want to use when declaring a target, eg. let's say to enable all warnings.

```python
C.cc_objects('Objects',
  sources = glob(join(project_dir, 'src/*.c')),
  additional_flags = [C.wall],
)
```

But yes, maybe something like this would be possible instead, but all flags would need to be duplicated on the `C` and `CXX` objects.

```python
C = load_module('compiler').C
C.objects('Objects',
  sources = glob(join(project_dir, 'src/*.c')),
  additional_flags = [C.wall],
)
```

**Comment by winksaville** [comment]
As long as setting the flags is obvious and easy seems reasonable. I assume
based on what you've written the syntax for other 'compiler's' would be:

Cxx = load_module('compiler').cxx
Asm = load_module('compiler').asm
....

Also, is it going to be "easy" for programmers to create there own
'compiler' modules and 'register" them with Creatr?

**Comment by winksaville** [comment]
Whoops Craftr :)

**Comment by NiklasRosenstein** [comment]
Almost, yes. Currently it looks like this:

```python
# craftr_module(hello_world)

P = load_module('platform')
Cxx = load_module('compiler').CxxCompiler()
Cxx.detect()

build_dir = join(project_dir, 'build')

Cxx.objects(
  'Objects',
  sources = glob(join(project_dir, 'source/**/*.cpp')),
)

Cxx.executable(
  'Program',
  name = join(build_dir, 'main'),
  inputs = [Objects],
)
```

Implemented in c90c4a7.

__Todo__:
- Support for optimization levels and other important compiler settings (what flags do you think are especially commonly used?)
- ASM compiler interface

**Comment by winksaville** [comment]
The minimum would be: release (-O2), debug (-O0 -g), debugoptimized (-O2
-g) but critically the author needs to be able to remove any predefined
flags and provide exactly what the author requires. In Meson these are
"buildtype's" and one of them is "plain" which sets the "compiler" flags
globally always to empty when creating a compiler. I suggest Craftr should
allow the flags to be defined as each "compiler" is created and the syntax
should allow adding to the defaults or replace the defaults.

My two centavos :)
